x/mm2/keeper/keeper.go

```diff
@@ -0,0 +1,203 @@
+func (k *Keeper) SetLastUpdated(ctx sdk.Context, height uint64) error {
+	return k.lastUpdated.Set(ctx, height)
+}
+
+// GetLastUpdated gets the last block-height the market map was updated.
+func (k *Keeper) GetLastUpdated(ctx sdk.Context) (uint64, error) {
+	return k.lastUpdated.Get(ctx)
+}
+
+// GetMarket returns a market from the store by its currency pair string ID.
+func (k *Keeper) GetMarket(ctx sdk.Context, tickerStr string) (types.Market, error) {
+	return k.markets.Get(ctx, types.TickerString(tickerStr))
+}
+
+// GetAllMarkets returns the set of Market objects currently stored in state.
+func (k *Keeper) GetAllMarkets(ctx sdk.Context) ([]types.Market, error) {
+	iter, err := k.markets.Iterate(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	markets, err := iter.Values()
+	if err != nil {
+		return nil, err
+	}
+	return markets, err
+}
+
+// GetAllMarketsMap returns the set of Market objects currently stored in state
+// as a map[TickerString] -> Markets.
+func (k *Keeper) GetAllMarketsMap(ctx sdk.Context) (map[string]types.Market, error) {
+	iter, err := k.markets.Iterate(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	keyValues, err := iter.KeyValues()
+	if err != nil {
+		return nil, err
+	}
+
+	m := make(map[string]types.Market, len(keyValues))
+	for _, keyValue := range keyValues {
+		m[string(keyValue.Key)] = keyValue.Value
+	}
+
+	return m, nil
+}
+
+// createMarket initializes a new Market.
+// The Ticker.String corresponds to a market, and must be unique.
+func (k *Keeper) createMarket(ctx sdk.Context, market types.Market) error {
+	// Check if Ticker already exists for the provider
+	alreadyExists, err := k.markets.Has(ctx, types.TickerString(market.Ticker.String()))
+	if err != nil {
+		return err
+	}
+	if alreadyExists {
+		return types.NewMarketAlreadyExistsError(types.TickerString(market.Ticker.String()))
+	}
+	// Create the config
+	return k.markets.Set(ctx, types.TickerString(market.Ticker.String()), market)
+}
+
+// updateMarket updates a Market.
+// The Ticker.String corresponds to a market, and exist unique.
+func (k *Keeper) updateMarket(ctx sdk.Context, market types.Market) error {
+	// Check if Ticker already exists for the provider
+	alreadyExists, err := k.markets.Has(ctx, types.TickerString(market.Ticker.String()))
+	if err != nil {
+		return err
+	}
+	if !alreadyExists {
+		return types.NewMarketDoesNotExistsError(types.TickerString(market.Ticker.String()))
+	}
+	// Create the config
+	return k.markets.Set(ctx, types.TickerString(market.Ticker.String()), market)
+}
+
+// CreateMarkets sets the market data for a given set of markets and validates the state  It also
+// sets the LastUpdated field to the current block height.
+func (k *Keeper) CreateMarkets(ctx sdk.Context, markets []types.Market) error {
+	for _, market := range markets {
+		if err := k.createMarket(ctx, market); err != nil {
+			return err
+		}
+	}
+
+	if err := k.validateState(ctx, markets); err != nil {
+		return err
+	}
```

Comment:
I understand the need to do validation after all updates have been made, but won't the fact that markets have been created already leave us in a weird state? We'll have created a bunch of markets and errored out but the last updated height won't be changed and we'll have no way to roll back to the previous valid state.

Reply:
No - this is a transaction. So the update to state is all or nothing. There won't be any weird dangling intermediate states